storage/postgresql/log_storage.go

```diff
@@ -0,0 +1,795 @@
+// Copyright 2016 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package postgresql
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/monitoring"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/storage/cache"
+	"github.com/google/trillian/storage/tree"
+	"github.com/google/trillian/types"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/transparency-dev/merkle/compact"
+	"github.com/transparency-dev/merkle/rfc6962"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"k8s.io/klog/v2"
+)
+
+const (
+	createTempQueueLeavesTable = "CREATE TEMP TABLE TempQueueLeaves (" +
+		" TreeId BIGINT," +
+		" LeafIdentityHash BYTEA," +
+		" LeafValue BYTEA," +
+		" ExtraData BYTEA," +
+		" MerkleLeafHash BYTEA," +
+		" QueueTimestampNanos BIGINT," +
+		" QueueID BYTEA," +
+		" IsDuplicate BOOLEAN DEFAULT FALSE" +
+		") ON COMMIT DROP"
+	queueLeavesSQL = "SELECT * FROM queue_leaves()"
```

Comment:
Given enough time I can probably chip away at the layers here to understand what's going on. But it strikes me that what would be super handy is a high level guide in a README.md in this directory that lays out the approach you're taking. If someone were to come and look at this code, how should they start to understand what goes in the temporary tables, the real tables, or the stored procedures? Are the temporary tables bound to a single transaction? Why are stored procedures used rather than direct queries?

I can guess answers to the above with a reasonable degree of accuracy I expect. But I'd feel much more at home if I didn't have to guess 😄  I don't expect this to be a massive tome. Just a couple of paragraphs to explain the approach and perspective you had in mind when this was written will be ace.

Reply:
Addressed by https://github.com/google/trillian/pull/3644/commits/8c40e33a7ad92cf11b83af87e2d2d405785036b6.

Reply:
Cheers!